### PR TITLE
Implement function macros

### PR DESCRIPTION
**Pull request by jyn514**

### CODE REVIEW

File: src/lex/cpp.rs [hunk 1]
```diff
@@ -691,28 +784,109 @@ impl<'a> PreProcessor<'a> {
         }
         Ok(())
     }
+    // before:
+    // #define f(a, b, c) a + b + c
+    //           ^
+    // after:
+    // #define f(a, b, c) a + b + c
+    //                   ^
```
Comment (pythondude325) [inline src/lex/cpp.rs]: Can you give some context to this comment?

File: src/lex/cpp.rs [hunk 2]
```diff
@@ -484,7 +512,72 @@ impl<'a> PreProcessor<'a> {
...
+        let mut current_arg = Vec::new();
+        // now, expand all arguments
+        loop {
+            let next = match self.next_replacement_token() {
+                None => return None,
+                Some(Err(err)) => return Some(Err(err)),
+                Some(Ok(token)) => token,
+            };
+            if next.data == Token::Comma {
+                args.push(mem::replace(&mut current_arg, Vec::new()));
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/lex/cpp.rs]: `take` requires 1.40, I want to keep backwards compatible with 1.37 (as documented in the README)

File: src/lex/cpp.rs (hunk 1 above)
Comment (jyn514) [reply src/lex/cpp.rs]: This shows the tokens that will be consumed by this function.

File: src/lex/cpp.rs [hunk 3]
```diff
@@ -157,6 +165,29 @@ impl<'a> PreProcessor<'a> {
     fn lexer_mut(&mut self) -> &mut Lexer {
         self.includes.last_mut().unwrap_or(&mut self.first_lexer)
     }
+    /// Return the next token, taking into account replacements
+    fn next_replacement_token(&mut self) -> Option<CppResult<Token>> {
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/lex/cpp.rs]: I think the explicit return is more clear here, though.

[the remark this answers is not in the document]
Comment (pythondude325) [reply src/lex/cpp.rs]: Fair

File: src/lex/cpp.rs (hunk 1 above)
Comment (pythondude325) [reply src/lex/cpp.rs]: Can you write that with the comment? At a first look, i thought this was referring to an error annotation

**Comment by jyn514** [comment]
Addressed review comments.

**Pull request merged**